README.md

```diff
@@ -120,9 +138,127 @@ module.exports = function (fastify, options, next) {
+## Example Request
+
+The following example request object is referenced in the default method documentation.
+
+```HTTP
+POST / HTTP/1.1
+Host: localhost:3000
+Signature: keyId="test-key-a", algorithm="hs2019", headers="(request-target) (created) (expires) host digest content-type", signature="pQul5YFrqv76Zq2bE1kWjJfFGnTu0MwU7X7c8MWDswAI5V7dROqKbBWKUGcoysxujgTqkJo/Eg74x34o54hqRg==", created=1402170695, expires=1402170895
+Digest: sha-512=+PtokCNHosgo04ww4cNhd4yJxhMjLzWjDAKtKwQZDT4Ef9v/PrS/+BQLX4IX5dZkUMK/tQo7Uyc68RkhNyCZVg==
+Content-Type: application/json
+
+{"hello": "world"}
+```
+
+The documentation assumes the fastify-hmac plugin has been registered with the following options.
+
+```js
+  fastify.register(require('fastify-hmac'), {
+    sharedSecret: 'topSecret',
+    algorithmMap: {
+      hs2019: {
+        'test-key-a': 'sha512'
+      }
+    }
+  })
+```
+
+## extractSignature Method
+
+The `extractSignature` method returns the HMAC signature string found in a request Signature header constructed according to [IETF draft standards][1].
+
+This method is called with two parameters, the fastify `request` object and the plugin `options` object.
+
+This signature is compared to the calculated HMAC signature from `constructSignatureString` to validate message authenticity.
+
+
+Given the [Example Request](#example-request) this method will:
+1. Parse the `request` Signature header string
+2.  Return the signature property value: `"pQul5YFrqv76Zq2bE1kWjJfFGnTu0MwU7X7c8MWDswAI5V7dROqKbBWKUGcoysxujgTqkJo/Eg74x34o54hqRg=="`.
+
+## constructSignatureString Method
+
+The `constructSignatureString` method returns a calculated HMAC signature string. 
+
+This method is called with two parameters, the fastify `request` object and the plugin `options` object.
+
+This signature is compared to the HMAC signature string found in the request Signature header to validate message authenticity. 
+
+
+Given the [Example Request](#example-request) this method will:
+1. Call [`getAlgorithm`](#getAlgorithm-Method) from the plugin `options` object and use its return value along with the `sharedSecret` property from the plugin `options` object to create a new HMAC object
+2. Update the HMAC object with the signature content string obtained from `getMessage`
+   1. The `getMessage` function returns a formatted string according to [IETF draft standards][1] containing the Signature content listed under `headers` in the `request` Signature header.
+3. Encode the signature with the digest encoding string returned from [`getSignatureEncoding`](#getSignatureEncoding-Method) in the plugin `options` object.
+4. Return the calculated signature string of: `"pQul5YFrqv76Zq2bE1kWjJfFGnTu0MwU7X7c8MWDswAI5V7dROqKbBWKUGcoysxujgTqkJo/Eg74x34o54hqRg=="`
```

Comment:
I've done some more thinking on this. I don't really think we want to describe the internals step by step as much. If someone wants to see how the default methods work, they can go check out the code. I think this would be a little easier to follow if it was something like:

```suggestion
## constructSignatureString

`constructSignatureString(request, options)`

Constructs and returns the signature string that will be compared to the HMAC signature string found in the request.

Given the example request, the default will return the calculated signature string of:

`"pQul5YFrqv76Zq2bE1kWjJfFGnTu0MwU7X7c8MWDswAI5V7dROqKbBWKUGcoysxujgTqkJo/Eg74x34o54hqRg=="`
```

We can also link directly to the file where the implementation is as well. This is less descriptive for sure, but I think it would point people in the correct direction. Happy to hear other takes on this as well 👍

Reply:
I would say this is optional though. I dont really know if that is a better solution or not. Writing docs is hard 😖 

We can gradually build these out if we need to.

Reply:
I agree with James here on all counts (docs are hard!). I think we don't need to walk through how the methods work but rather just show how to use them with a code example of them being called.